### ISSUE TITLE
Benchmarking on CIFAR-10

### ISSUE THREAD

**Issue by abhinavagarwalla** [facebookresearch/swav#23]
Hi,

I wanted to benchmark SwAV on CIFAR-10.
Is there any recommended configuration for CIFAR-10? For eg:
-  The number of prototypes could be set to 50, 100 etc.
-  Since CIFAR-10 images are 32x32, multicrop can be avoided.

Also, do you plan to publish any pretrained model on CIFAR-10?

**Comment by mathildecaron31** [comment]
Hi, Thanks for your interest.

I have not run any SwAV pre-training on CIFAR-10 dataset and it is not on my roadmap.

I agree with both your suggestions. I think trying different number of prototypes (10 - 30 - 100 - 300 for example) is a good idea. Also, multicrop might indeed not be relevant in CIFAR-10 regime, thus you can start with 2x32 crops.

**Comment by carmete** [comment]
@abhinavagarwalla Have you been able to find a good setup for CIFAR-10?

**Comment by abhinavagarwalla** [comment]
Hi @carmete 

I initially faced an issue with the loss being stuck at ln(nmb_prototypes). I was able to get around this by changing the hyperparameters.
I finally used num_prototypes = 30, temperature=0.5, base_lr: 0.06. Using these parameters, the loss did go below ln(30) and the model converged. I could get 69% linear eval accuracy which is lower than competing methods.
I didn't do too much hyperparameter tuning, which I think could be a way to get better results.

**Comment by ye-yechen** [comment]
> Hi @carmete
> 
> I initially faced an issue with the loss being stuck at ln(nmb_prototypes). I was able to get around this by changing the hyperparameters.
> I finally used num_prototypes = 30, temperature=0.5, base_lr: 0.06. Using these parameters, the loss did go below ln(30) and the model converged. I could get 69% linear eval accuracy which is lower than competing methods.
> I didn't do too much hyperparameter tuning, which I think could be a way to get better results.

I also want to use the cifar-10 dataset to train swav model,but I dont know how to pass the dataset. In the source code,the author just pass the path of the imagenet,but the cifar has the different format from imagenet. Can you help me and show me your code? Thank you!

**Comment by carmete** [comment]
@ye-yechen you could adapt this code (https://github.com/facebookresearch/swav/blob/master/src/multicropdataset.py) to also train on CIFAR-10:

```python
...
from torch.utils.data import Dataset, Subset
...
class MultiCropDataset(Dataset):
    def __init__(
...
        super(MultiCropDataset, self).__init__()
        assert len(size_crops) == len(nmb_crops)
        assert len(min_scale_crops) == len(nmb_crops)
        assert len(max_scale_crops) == len(nmb_crops)
        if data_path == 'cifar10':
            self.data = datasets.CIFAR10(root='data/CIFAR-10', download=True)
        else:
            self.data = datasets.ImageFolder(data_path)

        if size_dataset >= 0:
            self.data = Subset(self.data, list(range(size_dataset)))
        self.return_index = return_index
...
    def __len__(self):
        return len(self.data)

    def __getitem__(self, index):
        image, _ = self.data[index]
        multi_crops = list(map(lambda trans: trans(image), self.trans))
        if self.return_index:
            return index, multi_crops
        return multi_crops
...
```
where `"data/CIFAR-10"` should be an existing folder or should be replaced with the path to your CIFAR-10 dataset.

**Comment by ye-yechen** [comment]
@carmete OK，I will try it. Thank you for your help. Best wishes!

**Comment by ye-yechen** [comment]
> Hi @carmete
> 
> I initially faced an issue with the loss being stuck at ln(nmb_prototypes). I was able to get around this by changing the hyperparameters.
> I finally used num_prototypes = 30, temperature=0.5, base_lr: 0.06. Using these parameters, the loss did go below ln(30) and the model converged. I could get 69% linear eval accuracy which is lower than competing methods.
> I didn't do too much hyperparameter tuning, which I think could be a way to get better results.

Hi,thanks for your help.I use your codes and train the model with the cifar10. But when I run the eval_linear.py I get a very low accuracy... I dont know whether the hyperparameters were set correctly.  May I trouble you to give me a guidance?Thanks a lot!

**Comment by ye-yechen** [comment]
> Hi @carmete
> 
> I initially faced an issue with the loss being stuck at ln(nmb_prototypes). I was able to get around this by changing the hyperparameters.
> I finally used num_prototypes = 30, temperature=0.5, base_lr: 0.06. Using these parameters, the loss did go below ln(30) and the model converged. I could get 69% linear eval accuracy which is lower than competing methods.
> I didn't do too much hyperparameter tuning, which I think could be a way to get better results.

@abhinavagarwalla Hi,thanks for your help.I use your hyperparameters and train the model with the cifar10. But when I run the eval_linear.py I get a very low accuracy... I dont know whether the hyperparameters were set correctly. May I trouble you to give me a guidance?Thanks a lot!

**Comment by abhinavagarwalla** [comment]
@ye-yechen, the mentioned parameters worked for me.
Also check the number of crops. CIFAR-10 images are 32x32, thus multicrop doesn't make sense.
Do ensure that the training loss goes below ln(nmb_prototypes).
I also modified the ResNet architecture to be be more suitable for CIFAR-10, by reducing the kernel size to 3 for first conv. If you aren't doing that, I would recommend to do that.
For eval_linear.py, some parameters might need to be adjusted like learning rate, batch size, etc. For me, changing hyperparameters here didn't change the results much. But, might be helpful for you.

**Comment by ye-yechen** [comment]
> @ye-yechen, the mentioned parameters worked for me.
> Also check the number of crops. CIFAR-10 images are 32x32, thus multicrop doesn't make sense.
> Do ensure that the training loss goes below ln(nmb_prototypes).
> I also modified the ResNet architecture to be be more suitable for CIFAR-10, by reducing the kernel size to 3 for first conv. If you aren't doing that, I would recommend to do that.
> For eval_linear.py, some parameters might need to be adjusted like learning rate, batch size, etc. For me, changing hyperparameters here didn't change the results much. But, might be helpful for you.

@abhinavagarwalla  Hi, Due to the multicrop doesn't make sense in cifar-10,so what transformations should be applied to the cifar10?And  I noticed that the loss decreases from ln(30)(about 3.4) to 2.3, I dont know if the loss was in a normal level.Unfortunately, I still got a low accuracy. I feel a bit broken...

**Comment by LeeDoYup** [comment]
Could you let me know any update, or reasons why this model fails to learn CIFAR datasets, even though they are much easier to learn?

**Comment by abhinavagarwalla** [comment]
Hi @ye-yechen,
I put up the modifications I did in this [repository](https://github.com/abhinavagarwalla/swav-cifar10). It's nothing special but I hope it is helpful.

@LeeDoYup  In my opinion, the approach should work in principle. I think it boils down to architectural choices around prototypes ie. how many prototypes should be there, how frequently should they be updated. etc. Also compared to ImageNet, image diversity in CIFAR datasets is considerably low. Moreover, the number of prototypes employed are generally more than the number of classes. For CIFAR-10, around 30-50-100 work well but not 10. This could be another reason. @mathildecaron31 might have a better intuition.

**Comment by ye-yechen** [comment]
@abhinavagarwalla Thanks a lot! I will learn this codes. Best wishes!

**Comment by Ciccios96** [comment]
@ye-yechen there are any updates about this benchmark? i'm trying to get a good loss with the model modified by @abhinavagarwalla but i'm stuck at 3.4012. If i try to do a SVM linear evaluation i get a very low accuracy.

**Comment by Hzzone** [comment]
I will try, post here later...

**Comment by Hzzone** [comment]
As described in paper: ```We observe that a strong entropy regularization (i.e. using a high ε)
generally leads to a trivial solution where all samples collapse into an unique representation and are all assigned uniformely to all prototypes. Hence, in practice we keep ε low.``` I also found that a lower epsilon helps a lot. plz try 0.01.

**Comment by Hzzone** [comment]
I also noticed that the bs is very important for online clustering, where I trained resnet18, bs 2048, and 8 2080ti on cifar10. It seems to work well. Maybe further tuning is needed which are the most annoyed.

**Comment by carmete** [comment]
When I was experimenting with SwAV on CIFAR, I believe the most important part was to reduce the Color Jitter strength to "s=0.5" and adjusting the normalisation statistics here https://github.com/facebookresearch/swav/blob/master/src/multicropdataset.py#L39.

You should have something like:
```python
color_transform = [get_color_distortion(s=0.5)]
mean = [0.4914, 0.4822, 0.4465]
std = [0.2470, 0.2435, 0.2616]
```

Also note that usually for CIFAR the first ResNet Conv has kernel size 3 and the first maxpool is omitted. Hope this helps :)

**Comment by engmubarak48** [comment]
@carmete You said results should be close to SOTA!, Have you trained on CIFAR and achieved that. if so, which CIFAR ? is it CIFAR10 or 100. and also if it worked for you, can you kindly summarize all the hyper-parameters that worked for you. Thank you. 

PS: I have tried the suggested hyper-parameters, though the accuracy is fairly good compare to the random model, I didn't get SOTA results? that is why I am curious about your comment.

**Comment by engmubarak48** [comment]
Also, I have a question when traininging the SWAV model for shorter epochs (i.e. about 200) and longer epochs (i.e. 500) on CIFAR100, I get almost close results (i.e. close accuracy in the fine-tuning stage). Does this mean training longer is not that helpful. 

PS: my swav loss goes far below the ln(nb_prototypes) and ln(actual_class).

**Comment by carmete** [comment]
@engmubarak48: I only experimented with CIFAR-10, where I trained upto around 90% accuracy after 800 epochs (80.8 after 100 epochs).

As mentioned above:
1) Reduce augmentation strength
2) Change kernel size and stride

Something like this should work:

```python
#!/bin/bash

port=$RANDOM

# Experiment parameters
ARCH="resnet18_cifar"
hidden_mlp=512
ARCH_name="resnet18"
DATASET_PATH="cifar10"

# Augmentation parameters 
min_scale_crops=0.14
max_scale_crops=1.0

# Training parameters 
nepochs=100
bs=256
LR_start=0.6 # This is double of linear scaling rule
LR_end=0.0006 # This is double of linear scaling rule
warmup_epochs=5
size_crops=32
nmb_crops=2
nmb_prototypes=100
queue_length=120
epoch_queue_starts=15
use_fp16=true
freeze_prototypes_niters=5005

# Experiment path
EXP_SUFFIX="_${ARCH_name}_${nepochs}ep_${nmb_prototypes}prot"
EXPERIMENT_PATH="./experiments/swav_cifar${EXP_SUFFIX}"
mkdir -p $EXPERIMENT_PATH

# Run 
python3 -m torch.distributed.launch --nproc_per_node=1 --master_port=$port main_swav.py \
--data_path $DATASET_PATH \
--arch ${ARCH} \
--epochs $nepochs \
--base_lr $LR_start \
--final_lr $LR_end \
--warmup_epochs $warmup_epochs \
--batch_size $bs \
--size_crops $size_crops \
--nmb_crops $nmb_crops \
--min_scale_crops $min_scale_crops \
--max_scale_crops $max_scale_crops \
--use_fp16 $use_fp16 \
--freeze_prototypes_niters $freeze_prototypes_niters \
--nmb_prototypes $nmb_prototypes \
--queue_length $queue_length \
--hidden_mlp $hidden_mlp \
--epoch_queue_starts $epoch_queue_starts \
--dump_path $EXPERIMENT_PATH

mkdir ${EXPERIMENT_PATH}_linear_eval

python3 -m torch.distributed.launch --nproc_per_node=1 --master_port=$port eval_linear.py \
--num_labels 10 \
--crop_size 32 \
--arch $ARCH \
--data_path $DATASET_PATH \
--pretrained ${EXPERIMENT_PATH}/checkpoint.pth.tar \
--dump_path ${EXPERIMENT_PATH}_linear_eval
```

I removed some lines specific to the exact experiment I was doing, so please excuse if any errors pop up, because I accidentally removed something else as well.

**Comment by pederismo** [comment]
> @engmubarak48: I only experimented with CIFAR-10, where I trained upto around 90% accuracy after 800 epochs (80.8 after 100 epochs with loss =1.89 for 100 prototypes).
>
Have you tried with other numbers of prototypes? Was 100 the value that gave you the best accuracy at the end?

**Comment by nimaous** [comment]
Hi All

thanks for all your comments. do you think that resizing CIFAR from 32*32 to 256*256 and then use multi-cropping might be a good idea?

**Comment by mathildecaron31** [comment]
Hi @nimaous 

To use multi-crop with CIFAR-10 I recommend checking out PAWS repository:
https://github.com/facebookresearch/suncet/blob/731547d727b8c94d06c08a7848b4955de3a70cea/src/paws_train.py#L104-L108

They indeed conducted experiments with PAWS method (semi-supervised method derived from SwAV) with multi-crop on CIFAR-10 dataset.

Hope that helps

**Comment by Hzzone** [comment]
@nimaous absolutely not. The image size is not the most key point for clustering. Instead, cifar10 or 100 are different from Imagenet: they are blurry, and the objects are usually located in the center of each image. However, I always found that the large image size would bring a few performance improvements.

**Comment by nimaous** [comment]
@mathildecaron31  
thanks for your mention. I saw that in earlier posts you mentioned that there is no need for multi-cropping for CIFAR. if so, should we only use global views (weak cropping) because in cluster assignment the global view is required?  on the other hand, using only global views might make the optimization task easier for the model and result in a worse learned representation.

**Comment by yqy2001** [comment]
> @engmubarak48: I only experimented with CIFAR-10, where I trained upto around 90% accuracy after 800 epochs (80.8 after 100 epochs with loss =1.89 for 100 prototypes).
> 
> As mentioned above it is crucial to:
> 
> 1. Reduce augmentation strength
> 2. Change kernel size and stride
> 
> Only then, something like this should work:
> 
> ```python
> #!/bin/bash
> 
> port=$RANDOM
> 
> # Experiment parameters
> ARCH="resnet18_cifar"
> hidden_mlp=512
> ARCH_name="resnet18"
> DATASET_PATH="cifar10"
> 
> # Augmentation parameters 
> min_scale_crops=0.14
> max_scale_crops=1.0
> 
> # Training parameters 
> nepochs=100
> bs=256
> LR_start=0.6 # This is double of linear scaling rule
> LR_end=0.0006 # This is double of linear scaling rule
> warmup_epochs=5
> size_crops=32
> nmb_crops=2
> nmb_prototypes=100
> queue_length=120
> epoch_queue_starts=15
> use_fp16=true
> freeze_prototypes_niters=5005
> 
> # Experiment path
> EXP_SUFFIX="_${ARCH_name}_${nepochs}ep_${nmb_prototypes}prot"
> EXPERIMENT_PATH="./experiments/swav_cifar${EXP_SUFFIX}"
> mkdir -p $EXPERIMENT_PATH
> 
> # Run 
> python3 -m torch.distributed.launch --nproc_per_node=1 --master_port=$port main_swav.py \
> --data_path $DATASET_PATH \
> --arch ${ARCH} \
> --epochs $nepochs \
> --base_lr $LR_start \
> --final_lr $LR_end \
> --warmup_epochs $warmup_epochs \
> --batch_size $bs \
> --size_crops $size_crops \
> --nmb_crops $nmb_crops \
> --min_scale_crops $min_scale_crops \
> --max_scale_crops $max_scale_crops \
> --use_fp16 $use_fp16 \
> --freeze_prototypes_niters $freeze_prototypes_niters \
> --nmb_prototypes $nmb_prototypes \
> --queue_length $queue_length \
> --hidden_mlp $hidden_mlp \
> --epoch_queue_starts $epoch_queue_starts \
> --dump_path $EXPERIMENT_PATH
> 
> mkdir ${EXPERIMENT_PATH}_linear_eval
> 
> python3 -m torch.distributed.launch --nproc_per_node=1 --master_port=$port eval_linear.py \
> --num_labels 10 \
> --crop_size 32 \
> --arch $ARCH \
> --data_path $DATASET_PATH \
> --pretrained ${EXPERIMENT_PATH}/checkpoint.pth.tar \
> --dump_path ${EXPERIMENT_PATH}_linear_eval
> ```
> 
> I removed some lines specific to the exact experiment I was doing, so please excuse if any errors pop up, because I accidentally removed something else as well.

@carmete Hi, I find that you use resnet18 here, can you kindly share the code of model structure for us to reproduce this result on cifar10? Thanks!

**Comment by YunYunY** [comment]
> Hi @carmete

Thank you for sharing the code. Could you please share under what setting, your model stuck? I am curious to know changing which parameter would cause the issue. 
Thank you. 
> 
> I initially faced an issue with the loss being stuck at ln(nmb_prototypes). I was able to get around this by changing the hyperparameters. I finally used num_prototypes = 30, temperature=0.5, base_lr: 0.06. Using these parameters, the loss did go below ln(30) and the model converged. I could get 69% linear eval accuracy which is lower than competing methods. I didn't do too much hyperparameter tuning, which I think could be a way to get better results.
